File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.js
// /dev/stdin
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [K, N] = input[0].split(" ").map(Number);
const lines = [];
for (let i = 1; i <= K; i++) {
  lines.push(Number(input[i]));
}
let answer = 0;

parametricSearch(lines);

function parametricSearch(lines) {
  let left = 0;
  let right = Math.max(...lines);

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    let count = 0;
    for (const line of lines) {
      count += Math.floor(line / mid);
    }
    if (count >= N) {
      answer = mid;
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
}

console.log(answer);

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
"""

"""
import sys
from heapq import *
                                  
K, N = map(int, sys.stdin.readline().strip().split(' '))
lan = [int(sys.stdin.readline().strip()) for _ in range(K)]
# print(lan)

def search(arr, target):
  left = 0
  right = max(arr)
  result = 0
  while left <= right:
    mid = (left + right) // 2
    tempArr = [x for x in arr]

    for i in range(len(tempArr)):
      if tempArr[i] != 0 and mid != 0:
        tempArr[i] = tempArr[i] // mid
    # tempArr = list(map(lambda x : x // mid, tempArr))
    # print(sum(tempArr), tempArr,mid)

    if sum(tempArr) < target:
      right = mid - 1
    else:
      left = mid + 1

  return right



print(search(lan, N))



  

